src/test/java/bio/terra/janitor/service/cleanup/FlightManagerTest.java

```diff
@@ -0,0 +1,182 @@
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+@Tag("unit")
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = Main.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_CLASS)
+public class FlightManagerTest {
+  private static final Instant CREATION = Instant.EPOCH;
+  private static final Instant EXPIRATION = CREATION.plusSeconds(60);
+
+  @Autowired StairwayComponent stairwayComponent;
+  @Autowired JanitorDao janitorDao;
+
+  private static TrackedResource newResourceForCleaning() {
+    return TrackedResource.builder()
+        .trackedResourceId(TrackedResourceId.create(UUID.randomUUID()))
+        .trackedResourceState(TrackedResourceState.READY)
+        .cloudResourceUid(
+            new CloudResourceUid()
+                .googleBucketUid(new GoogleBucketUid().bucketName(UUID.randomUUID().toString())))
+        .creation(CREATION)
+        .expiration(EXPIRATION)
+        .build();
+  }
+
+  private void blockUntilFlightComplete(String flightId)
+      throws InterruptedException, DatabaseOperationException {
+    Duration maxWait = Duration.ofSeconds(10);
+    Duration waited = Duration.ZERO;
+    while (waited.compareTo(maxWait) < 0) {
+      if (!stairwayComponent.get().getFlightState(flightId).isActive()) {
+        return;
+      }
+      int pollMs = 100;
+      waited.plus(Duration.ofMillis(pollMs));
+      TimeUnit.MILLISECONDS.sleep(pollMs);
+    }
+    throw new InterruptedException("Flight did not complete in time.");
+  }
+
+  /** Assert that the cleanup flight is finishing. */
+  private void assertFlightFinishing(String flightId) {
+    assertEquals(janitorDao.getFlightState(flightId), Optional.of(CleanupFlightState.FINISHING));
+  }
+
+  @Test
+  public void scheduleFlight() throws Exception {
+    FlightManager manager =
+        new FlightManager(
+            stairwayComponent.get(),
+            janitorDao,
+            trackedResource ->
+                FlightFactory.FlightSubmission.create(OkCleanupFlight.class, new FlightMap()));
+    TrackedResource resource = newResourceForCleaning();
+    janitorDao.createResource(resource, ImmutableMap.of());
+
+    Optional<String> flightId = manager.submitFlight(EXPIRATION);
+    assertTrue(flightId.isPresent());
+    blockUntilFlightComplete(flightId.get());
+
+    assertFlightFinishing(flightId.get());
+  }
+
+  @Test
+  public void scheduleFlight_nothingReady() {
+    // No resources for cleaning inserted.
+    FlightManager manager =
+        new FlightManager(
+            stairwayComponent.get(),
+            janitorDao,
+            trackedResource ->
+                FlightFactory.FlightSubmission.create(OkCleanupFlight.class, new FlightMap()));
+    assertFalse(manager.submitFlight(EXPIRATION).isPresent());
+  }
+
+  @Test
+  public void recoverUnsubmittedFlights_unsubmittedFlight() throws Exception {
+    FlightManager manager =
```

Comment:
ditto, we should make injectable

Reply:
ditto, don't want to expose FlightManager to the entire application.